spec/abstract/SchedContext_A.thy

```diff
@@ -254,14 +254,15 @@ where
      \<equiv> whileLoop (\<lambda>_ s. the ((head_insufficient sc_ptr) s))
                   (\<lambda>_. non_overlapping_merge_refills sc_ptr) ()"
 
+definition "MAX_RELEASE_TIME = max_time - 5 * MAX_PERIOD"
```

Comment:
If we are adding this notion this far down in the abstract spec, we should use lower case letters as a derived notion. Then we'd want to state the equality between that and `maxReleaseTime` in StateRelation.thy.

As an alternative, we could move it to MachineExports.thy or the like and have the equality added in Hardware_H.

I can see that, on the haskell side, we are missing the `MaxPeriod` alias and always going back to `MaxPeriodUs`. It would make sense to add and use that, too.

Reply:
OK. Does that mean that `MAX_PERIOD` should be lowercase too? It's derived from `MAX_PERIOD_US` and `us_to_ticks`.

Reply:
What I meant is that this is not a good place to define a C level constant, which all upper case naming suggests. We could either move this definition to an appropriate place, or use lower case to consider it as an abstract spec constant. But either way we probably should maintain that distinction. `MAX_PERIOD` is defined in MachineOps. So, that makes sense.
Also, even if we are to choose using lower case, we'd better move the definition to Structure_A or somewhere anyway.

Reply:
I agree with Miki, if this can be moved to MachineOps and used by both specs then that would be nice. We're generally keen to reuse more things when possible instead of defining them twice.

Reply:
`MachineOps` is an arch-dependent file. If I were to move `MAX_RELEASE_TIME` there, we'd have to do it for all architectures. And if we wanted to modify the definition (say, changing the `5` to a `6`), we'd need to update it in all files. Currently, `MAX_RELEASE_TIME` is using the appropriate arch dependent constant, namely `MAX_PERIOD`, in both architectures. Is this not ideal?

Is there a convention again having constant being defined in certain files?

And while we're discussing this constant, I actually think it's a bad name. It is not an upper bound on the release time of a scheduling context: `refill_budget_check` and `refill_update` are both able to push refills past this point. It would be better to have it called `OVERRUN_LOOP_CAP` or something like that, but I know that the systems people won't want to change the name.

Reply:
Whether this constant should be arch specific or not is a separate question... this is not a good place to define it either way. We can assume that it is generic for now and define it maybe in MachineExports.thy as I suggested in the above.

After all, Refine isn't arch-split yet anyway.

I don't think we want to be dealing with different names from C for constants coming from C. You are welcome to challenge the naming with the systems team.